### PR TITLE
Implement tracing for Wan on BH Quad

### PR DESCRIPTION
**Pull request by sadesoyeTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/36032

### Problem description
No tracing for wan2.2 which is essential for multihost mesh.

### What's changed

- Pulled in new tracing class from [@friedrich's implementation](https://github.com/tenstorrent/tt-metal/tree/friedrich/tracing-wan)
- Added function decorator to make reuse easy.
- Fix wan's tracing integration on quad multihost.

### Checklist

- [ ] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:sadesoye/friedrich_tracing-wan)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:sadesoye/friedrich_tracing-wan)

#### Model tests

[T3K Integration Tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-integration-tests.yaml?query=branch%3Asadesoye%2Ffriedrich_tracing-wan)
[T3K Unit Tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml?query=branch%3Asadesoye%2Ffriedrich_tracing-wan) -Unrelated failure. Same on main.
[T3K Performance Tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-perf-tests.yaml?query=branch%3Asadesoye%2Ffriedrich_tracing-wan)
[Galaxy Integration Tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-integration-tests.yaml?query=branch%3Asadesoye%2Ffriedrich_tracing-wan) -Test passes, not sure why failure is reported.
[Galaxy Performance Tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-perf-tests.yaml?query=branch%3Asadesoye%2Ffriedrich_tracing-wan)
[Blackhole Demo Tests](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-demo-tests.yaml?query=branch%3Asadesoye%2Ffriedrich_tracing-wan)

### CI Status
_Auto-generated on every push. Click a pipeline name to open the workflow dispatch page._

| Pipeline | Status | Latest Run |
|---|:---:|---|
| [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml) | ✅ | [#1915](https://github.com/tenstorrent/tt-metal/actions/runs/23954471950) |
| [Blackhole post-commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) | ✅ | [#17046](https://github.com/tenstorrent/tt-metal/actions/runs/23954726142) |
| [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml) | ➖ | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml) |
| [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml) *(opt-in)* | ➖ | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml) |
| [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml) *(opt-in)* | ➖ | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml) |
| [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml) *(opt-in)* | ➖ | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml) |

### CODE REVIEW

File: models/tt_dit/tests/models/wan2_2/test_performance_wan.py [hunk 1]
```diff
@@ -100,19 +102,30 @@ def wan_pipeline_metrics_condimg(mesh_device, width, height, model_type):
...
+    config = ttnn.FabricRouterConfig()
+    config.max_packet_payload_size_bytes = 8192
+    return config
+
+
+device_params_8k = ring_params
+if is_blackhole():
+    device_params_8k.update({"fabric_router_config": create_fabric_router_config()})
+
+
```
[the remark this answers is not in the document]
Comment (sadesoyeTT) [reply models/tt_dit/tests/models/wan2_2/test_performance_wan.py]: That is why there is a is_blackhole() check in there.

**Review by sosborne-TT — APPROVED** [review]

File: models/tt_dit/pipelines/wan/pipeline_wan.py [hunk 2]
```diff
@@ -291,6 +314,31 @@ def __init__(
             1, self.vae.config.z_dim, 1, 1, 1
         )
 
+        # TODO: Reset buffers for change in resolution. Also reinitialize trace
```
Comment (cglagovichTT) [inline models/tt_dit/pipelines/wan/pipeline_wan.py:325]: still todo?

File: models/tt_dit/tests/models/wan2_2/test_performance_wan.py [hunk 3]
```diff
@@ -100,6 +102,17 @@ def wan_pipeline_metrics_condimg(mesh_device, width, height, model_type):
     return pipeline_cls, image_prompt, expected_metrics
 
 
+def create_fabric_router_config():
```
Comment (cglagovichTT) [inline models/tt_dit/tests/models/wan2_2/test_performance_wan.py]: could we put this 8k packet size router config in a utility file?

File: models/tt_dit/pipelines/wan/pipeline_wan.py (hunk 2 above)
Comment (sadesoyeTT) [reply models/tt_dit/pipelines/wan/pipeline_wan.py:325]: Yes. Right now it assumes 720p for GLX and 480p for QB/LB. I'll fix in another PR.

**Comment by jvegaTT** [comment]
Please give the PR a proper name

File: models/tt_dit/models/transformers/wan2_2/transformer_wan.py [hunk 4]
```diff
@@ -4,12 +4,17 @@
...
+import os
 from pathlib import Path
 
 import torch
 from diffusers.models.transformers.transformer_wan import WanRotaryPosEmbed as TorchWanRotaryPosEmbed
 from loguru import logger
 
+if not os.environ.get("TT_DIT_DEBUG"):
+    logger.disable(__name__)
+
```
Comment (jvegaTT) [inline models/tt_dit/models/transformers/wan2_2/transformer_wan.py:16]: Do you want to keep this debug stuff in the release code?

File: models/tt_dit/models/transformers/wan2_2/transformer_wan.py [hunk 5]
```diff
@@ -22,7 +27,8 @@
 from ....utils.mochi import get_rot_transformation_mat
 from ....utils.padding import pad_vision_seq_parallel
 from ....utils.substate import pop_substate, rename_substate
-from ....utils.tensor import bf16_tensor, float32_tensor, from_torch, unflatten
+from ....utils.tensor import bf16_tensor, from_torch, unflatten
+from ....utils.tracing import traced_function
```
Comment (amorrisonTT) [inline models/tt_dit/models/transformers/wan2_2/transformer_wan.py:31]: Most python linters would ban these nested relative imports

File: models/tt_dit/models/transformers/wan2_2/transformer_wan.py (hunk 4 above)
Comment (sadesoyeTT) [reply models/tt_dit/models/transformers/wan2_2/transformer_wan.py:16]: Yes. that's why it was added to prevent dumping some of the debug logs. We'll eventually clean this up as well as the logs once things are finalized.

**Comment by sadesoyeTT** [comment]
> Please give the PR a proper name

Totally forgot. Done.

File: models/tt_dit/models/transformers/wan2_2/transformer_wan.py (hunk 5 above)
Comment (sadesoyeTT) [reply models/tt_dit/models/transformers/wan2_2/transformer_wan.py:31]: Yea our linter seems to be fine with it. Its pretty much all over the codebase at this point.

**Review by amorrisonTT — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]

File: models/tt_dit/tests/models/wan2_2/test_performance_wan.py [hunk 6]
```diff
@@ -215,15 +218,17 @@ def test_pipeline_performance(
...
-                image_prompt=image_prompt,
-                height=height,
-                width=width,
-                num_frames=num_frames,
-                num_inference_steps=2,  # Small number of steps to reduce test time.
-            )
+        if traced:
+            with torch.no_grad():
+                pipeline(
+                    prompt=prompts[0],
```
[the remark this answers is not in the document]
Comment (cglagovichTT) [reply models/tt_dit/tests/models/wan2_2/test_performance_wan.py:224]: @sadesoyeTT is this correct? Or since pipeline init runs a warmup run, untraced perf time will be correct?

**Review by cglagovichTT — APPROVED** [review]

File: models/tt_dit/tests/models/wan2_2/test_performance_wan.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sadesoyeTT) [reply models/tt_dit/tests/models/wan2_2/test_performance_wan.py:224]: Yes, its the later, copilot failed to connect the dot. It flagged that init always calls warmup, and then flag this. Having the initial warmup in the constructor helps easily add an additional trace warmup later here and in the inference server.
